### PR TITLE
Make it more clear which groups a user is part of

### LINKED ISSUE #20282: Allow non-admin users to see members of their groups
I have set up multiple groups (e.g. HR, Developers, Management etc.) for my users so that they can share files with each other easily.

Unfortunately, it seems very difficult for non-admin user to see the groups they are in, especially when sharing files. I have only just found out that the groups users are part of are shown under their 'Personal' page. Here are my suggestions:
- [ ] Ideally, there should be another page called 'Groups' that shows which groups a member is a part of, and who is in those groups.

As a **quick fix**, the following could be done:
- [ ] On the 'Personal' page there be an anchor on the left hand side to 'Groups'
- [ ] In the 'Groups' section, all users who are members of each group should be listed

**sharing files**
- [ ] when clicking the 'share' button in the files app there should be a link showing all the user's groups, or ideally a list of all users (not sure if this is advisable e.g. when using LDAP)
- [ ] Perhaps when hovering over the 'i' it could say "You can share with people in these groups..."

### PR DESCRIPTION
**Pull request by pjrobertson**
Fixes #20282 

This is a first take at fixing #20282. Some of the ideas are controversial, so perhaps they need more discussion, but I wanted to get the ball rolling

Edit: I feel the 1st commit is harmless, the 2nd may need more discussion. Please see #20282 

Disclaimer: (Just seen @nickvergessen 's note) - I'm in no way a PHP developer, so take my code with a pinch of salt. This is more a proof of concept. Thanks!

### CODE REVIEW

**Comment by MorrisJobke** [comment]
cc @LukasReschke @jancborchardt @owncloud/designers

File: settings/personal.php [hunk 1]
```diff
@@ -146,9 +146,13 @@
...
 
 // Get array of group ids for this user
-$groups = \OC::$server->getGroupManager()->getUserIdGroups(OC_User::getUser());
-$groups2 = array_map(function($group) { return $group->getGID(); }, $groups);
-sort($groups2);
+$groupManager = \OC::$server->getGroupManager();
+$groups = $groupManager->getUserIdGroups(OC_User::getUser());
+$groups2 = array();
+foreach ($groups as $group) {
+    $groups2[$group->getGID()] =  $groupManager->displayNamesInGroups([$group]);
```
Comment (nickvergessen) [inline settings/personal.php]: This will explode on huge installations.

Comment (LukasReschke) [inline settings/personal.php]: :+1: - I'd rather avoid this thus…

File: settings/templates/personal.php [hunk 2]
```diff
@@ -144,7 +144,9 @@
 	<h2><?php p($l->t('Groups')); ?></h2>
 	<p><?php p($l->t('You are member of the following groups:')); ?></p>
 	<p>
-	<?php p(implode(', ', $_['groups'])); ?>
+	<?php foreach($_['groups'] as $groupName => $groupUsers) {
+        p("<strong>$groupName</strong>");
```
Comment (LukasReschke) [inline settings/templates/personal.php]: `p()` will also sanitize `<strong>` here

**Comment by rullzer** [comment]
:-1: from me. For the already mentioned issues. If you have large installations this will explode both in terms of performance and in terms of data to display.

You first patch (in isolation) would make sense I think.

**Comment by MorrisJobke** [comment]
> You first patch (in isolation) would make sense I think.

https://github.com/owncloud/core/pull/20979 cherry picked version of this.

Regarding the group member listing I also vote against this: :-1: it will simply not scale on big installations.

**Comment by pjrobertson** [comment]
Ok, I accept your reasoning. There is still a usability issue here though:
if I am a non-admin user, how do I know which group has which users?

E.g. I know I want to share with the "marketing" department, but I'm not
sure if that includes "Bill". How can I check?

I've had users run to me countless times with these simple kinds of
questions.

What does "not scale" mean exactly? Isn't the same argument then also
applocable for the admin "Users" section?

**Comment by jancborchardt** [comment]
We could do it similarly as Github does. When you hover over a group they display the members in a tooltip.

**Comment by rullzer** [comment]
@pjrobertson somewhat true. However the number of admin users is very limited and there we have some pagination etc.

@jancborchardt well the same argument holds there. Assume a university where everybody is a member of the group students (for example). if you would hoven over that you would fill several screens.

So I agree we should come up with a solution for this but simply listing all users is not the solution I think. Becaus then also the usability is gone once you get groups with more than 20 users.

**Comment by jancborchardt** [comment]
For the tooltip it should for sure only list the first ~10 people and then say »and 23 more …« for example.

**Comment by nickvergessen** [comment]
Which then defeats the requested functionality and is of no use again...

**Comment by jancborchardt** [comment]
It’s one step in the right direction at least. Currently there is no indication at all.

**Comment by nickvergessen** [comment]
Btw on LDAP we can't even count users, so this becomes quite messy

**Comment by jancborchardt** [comment]
Then list the first 10 and write »and more …« without listing the exact count. Or don’t show anything on LDAP. Not being able to do it perfectly in every environment is not an argument against the whole thing.

**Pull request closed without merging**